node/src/components/linear_chain_sync.rs

```diff
@@ -63,23 +68,68 @@ pub(crate) struct LinearChainSync<I> {
     /// we need to shut down for an upgrade.
     next_upgrade_activation_point: Option<ActivationPoint>,
     stop_for_upgrade: bool,
+    /// Key for storing the linear chain sync state.
+    state_key: Vec<u8>,
 }
 
 impl<I: Clone + PartialEq + 'static> LinearChainSync<I> {
     pub fn new(
         registry: &Registry,
+        chainspec: &Chainspec,
         init_hash: Option<BlockHash>,
         genesis_validator_weights: BTreeMap<PublicKey, U512>,
     ) -> Result<Self, prometheus::Error> {
         let state = init_hash.map_or(State::None, |init_hash| {
             State::sync_trusted_hash(init_hash, genesis_validator_weights)
         });
+        let state_key = state_key(&chainspec);
+        Ok(LinearChainSync {
+            peers: PeersState::new(),
+            state,
+            metrics: LinearChainSyncMetrics::new(registry)?,
+            next_upgrade_activation_point: None,
+            stop_for_upgrade: false,
+            state_key,
+        })
+    }
+
+    /// Initialize `LinearChainSync` component from preloaded `State`.
+    pub fn from_state(
+        registry: &Registry,
+        chainspec: &Chainspec,
+        state: State,
+    ) -> Result<Self, prometheus::Error> {
+        let state_key = state_key(&chainspec);
+        let state_status = match &state {
+            State::SyncingTrustedHash {
+                trusted_hash,
+                highest_block_seen,
+                ..
+            } => format!(
+                "syncing trusted hash: {:?}, highest block seen: {:?}",
+                trusted_hash, highest_block_seen
+            ),
+            State::SyncingDescendants {
+                trusted_hash,
+                latest_block,
+                ..
+            } => format!(
+                "syncing descendants of {:?}, latest block hash={:?}, height={:?}, era={:?}",
+                trusted_hash,
+                latest_block.hash(),
+                latest_block.height(),
+                latest_block.era_id(),
+            ),
+            _ => panic!("unexpected initial state: {:?}", &state),
+        };
+        info!(?state_status, "reusing previous state");
```

Comment:
Would it be cleaner to `impl Display for State`?
```suggestion
        info!(state_status = %state, "reusing previous state");
```

Reply:
Good idea.

Reply:
Done in [`6d6256a` (#904)](https://github.com/CasperLabs/casper-node/pull/904/commits/6d6256afbfb52e3c3bcc4ea817550a3ce25cea36).